src/EFCore.Analyzers/InternalUsageDiagnosticAnalyzer.cs

```diff
@@ -55,41 +55,48 @@ public override void Initialize(AnalysisContext context)
     {
-        switch (context.Operation.Kind)
+        switch (context.Operation)
         {
-            case OperationKind.FieldReference:
-                AnalyzeMember(context, ((IFieldReferenceOperation)context.Operation).Field);
+            case IFieldReferenceOperation fieldReference:
+                AnalyzeMember(context, fieldReference.Field);
                 break;
-            case OperationKind.PropertyReference:
-                AnalyzeMember(context, ((IPropertyReferenceOperation)context.Operation).Property);
+
+            case IPropertyReferenceOperation propertyReference:
+                AnalyzeMember(context, propertyReference.Property);
                 break;
-            case OperationKind.EventReference:
-                AnalyzeMember(context, ((IEventReferenceOperation)context.Operation).Event);
+
+            case IEventReferenceOperation eventReference:
+                AnalyzeMember(context, eventReference.Event);
                 break;
-            case OperationKind.MethodReference:
-                AnalyzeMember(context, ((IMethodReferenceOperation)context.Operation).Method);
+
+            case IMethodReferenceOperation methodReference:
+                AnalyzeMember(context, methodReference.Method);
                 break;
-            case OperationKind.ObjectCreation when ((IObjectCreationOperation)context.Operation).Constructor is { } constructor:
+
+            case IObjectCreationOperation { Constructor: { } constructor }:
                 AnalyzeMember(context, constructor);
                 break;
-            case OperationKind.Invocation:
-                AnalyzeInvocation(context, (IInvocationOperation)context.Operation);
+
+            case IInvocationOperation invocation:
+                AnalyzeInvocation(context, invocation);
                 break;
-            case OperationKind.VariableDeclaration:
-                AnalyzeVariableDeclaration(context, ((IVariableDeclarationOperation)context.Operation));
+
+            case IVariableDeclarationOperation variableDeclaration:
+                AnalyzeVariableDeclaration(context, variableDeclaration);
                 break;
-            case OperationKind.TypeOf:
-                AnalyzeTypeof(context, ((ITypeOfOperation)context.Operation));
+
+            case ITypeOfOperation typeOf:
+                AnalyzeTypeof(context, typeOf);
                 break;
+
             default:
-                throw new ArgumentException($"Unexpected {nameof(OperationKind)}: {context.Operation.Kind}");
+                throw new ArgumentException($"Unexpected operation: {context.Operation.Kind}");
         }
     }
 
     private static void AnalyzeMember(OperationAnalysisContext context, ISymbol symbol)
     {
-        // ReSharper disable once RedundantCast
-        if ((object)symbol.ContainingAssembly == context.Compilation.Assembly)
+        if (ReferenceEquals(symbol.ContainingAssembly, context.Compilation.Assembly))
```

Comment:
I believe the right way to compare symbols is this (also below):

```suggestion
        if (symbol.ContainingAssembly.Equals(context.Compilation.Assembly, SymbolEqualityComparer.Default))
```

Reply:
Yes, but the previous code looks like it was specifically trying to compare by reference. I'll change all occurrences to proper comparison if you are sure that previous behavior wasn't intentional.

Reply:
I don't think there was anything intentional here about comparing by reference - this is old code that may even have been written before SymbolEqualityComparer was a thing. In any case, we have tests which should tell us if the behavior breaks as a resuilt - so let's go ahead and do this.